### ISSUE TITLE
[Dynamic Fuzzing] Adversarial & Perturbation Input Trigger Generator

### ISSUE THREAD

**Issue by KrishG7** [KrishG7/Aegis-Tensor#6]
### Assignee: @7dxksh7
**Subsystem:** Dynamic Fuzzing & Activation Hook Engine  
**Sprint Due Date:** October 20, 2026 (Milestone 3)

---

### 1. Problem & Motivation
Sleeper Agent Trojans remain completely dormant under benign evaluation inputs. To surface backdoors without access to the attacker's secret trigger, we must dynamically fuzz the model across diverse adversarial distribution boundaries (e.g. extreme magnitude noise, frequency masks, patch triggers).

---

### 2. Implementation Steps
1. **Implement Input Perturbation Generators in `aegis/fuzzer.py`:**
   - `generate_gaussian_noise(shape, sigma=1.0)`: Perturbations covering wide variance ranges.
   - `generate_boundary_inputs(shape)`: Tests extreme inputs (clamped 0.0$, 569X10.0$, sparse delta impulses).
   - `generate_patch_triggers(shape, patch_size=(8, 8))`: Injects high-contrast geometric watermarks/squares simulating BadNets vision backdoors.
   - `generate_token_perturbations(vocab_size, seq_len)`: Extreme token repetition and out-of-distribution embeddings for language models.

2. **Integrate into `DynamicTrojanFuzzer`:**
   - Provide an automated fuzzing mode where the user does not need to supply their own generator function.

---

### 3. How to Test & Verify
Run the test suite with pytest:
```bash
source .venv/bin/activate
pip install torch
pytest -v tests/test_fuzzer.py
```

**Executable Verification Script:**
Save and run `scratch/verify_trigger_fuzzing.py`:
```python
import torch
import torch.nn as nn
from aegis.fuzzer import DynamicTrojanFuzzer

# Construct a model with an artificial sleeper backdoor
class TrojanModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(10, 20)
        self.fc2 = nn.Linear(20, 2)
    def forward(self, x):
        # Trigger condition: first element > 8.0
        h = self.fc1(x)
        if (x[:, 0] > 8.0).any():
            h = h * 50.0  # Massive artificial spike
        return self.fc2(h)

model = TrojanModel()
fuzzer = DynamicTrojanFuzzer(model, spike_threshold=4.0)

# Fuzz with boundary values
report = fuzzer.run_fuzzing(
    sample_inputs_generator=lambda i: torch.randn(2, 10) * (i + 1),
    baseline_input=torch.zeros(2, 10),
    num_iterations=20
)

print('Trojan detected:', report.suspected_trojan)
print('Spike ratio:', report.spike_ratio)
assert report.suspected_trojan == True
assert 'fc1' in report.suspicious_layers
print('TEST PASSED: Fuzzer successfully uncovered Trojan trigger!')
```
